README.md
# Popular Stream

[![Build Status](https://travis-ci.org/nhocki/popular_streams.svg?branch=master)](https://travis-ci.org/nhocki/popular_streams)

Simple popular content tracker with a Redis backend.

Mostly taken from the "Popular Stream" code found in the
[Surfacing Interesting Content](http://stdout.heyzap.com/2013/04/08/surfacing-interesting-content/)
post but bundled up as a gem.

PopularStream tracks an "event" on a group of "fields" and returns the ones that
are currently popular.

For example, if you want to track the most popular tags right now you, the event
would be "tagging" and the field would be the tag name ("rubygems" for example).

The way this works is that "old" votes will count less than newer votes, that way
a tag that was used 20 times today will be more popular than a tag used 30 times
last week.

[Here are more details about how this gem works.](http://blog.nhocki.com/2014/12/02/trending-content-with-ruby-and-redis/)

This gem was extracted from the [We Heart It](http://weheartit.com) codebase.

Once again, thanks to Micah Fivecoate for his post.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'popular_stream'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install popular_stream

This gems requires **Ruby 2.1**.

## Usage

There are two main methods `vote` and `get`. `vote` adds an event to the list
and `get` gets the most popular fields on the distribution.

```ruby
stream = PopularStream.new("popular_tags")

stream.vote(field: 'rubygems')

# You can also add an optional `weight` param.
stream.vote(field: 'ruby', weight: 2)

# And you can even specify when the event happened.
# Notice that time is a number, meaning seconds since Epoc.
time = Date.yesterday.to_time
stream.vote(field: 'rubygems', time: time.to_i)

stream.get # => ['rubygems', 'ruby']

# You can pass `limit` and `offset`
stream.get(limit: 1, offset: 1) # => ['ruby']
stream.get(offset: 10) # => []

# You can ask to get the scores as well
stream.get(with_scores: true) # => [['ruby', 0.00001]]
```

## Seting up & Configuring Redis

Popular Stream uses Redis as the storage database. By default it will connect to
the redis client on `ENV['REDIS_URL']`. You can also specify what redis client to
use with `PopularStream.redis = Redis.new(host: 'example.com')`.

Notice that this is for *all* streams. This is because we don't want to create new
connections for every new stream instance we create, so the same client is used.

## TODO

* Sinatra application. I want to bundle an optional Sinatra application that
makes this gem super simple to setup as a service.

* Multiple storage databases. Right now everything's stored on Redis but it should
be simple to use more stuff. People should be able to create "adapters" and use
them as needed.

## Contributing

1. Fork it ( https://github.com/nhocki/popular_stream/fork )
2. Create your a new branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

If you're gonna open a Pull Request, please install [EditorConfig](http://editorconfig.org/),
it will keep most of the basic code styling consistent.


CHANGELOG.md
# Version 1.1 (Nov. 30, 2014)

* Adding ability to get the scores with the results. This will be the default
in V 2.0.0

# Version 1.0 (Nov. 29, 2014)

* Extracting code from We Heart It and making it a gem.



spec/spec_helper.rb
require 'popular_stream'
require 'fakeredis/rspec'

Bundler.require(:default, :test, :development)

RSpec.configure do |config|
  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
    expectations.syntax = :expect
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  config.disable_monkey_patching!
end


spec/lib/popular_stream_spec.rb
require "spec_helper"

RSpec.describe PopularStream do
  let(:one_week) { 60 * 60 * 24 * 7 } # 1 week in seconds.
  let(:stream) { PopularStream.new("popular_stream") }

  # Make a new "Redis" for each test to not leak the doubles.
  before do
    PopularStream.redis = Redis.new
    stream.clear!
  end

  describe ".redis" do
    let(:redis) { double }

    it "is possible to set a redis instace" do
      expect { PopularStream.redis = redis }.not_to raise_error
    end
  end

  describe "#get" do
    before do
      stream.vote(field: '1', time: Time.now - one_week)
      stream.vote(field: '2')
    end

    it "ranks older votes lower than newer votes" do
      expect(stream.get).to eql(['2', '1'])
    end

    it "#get takes limit & offset arguments" do
      expect(stream.get(limit: 1)).to eql(['2'])
      expect(stream.get(offset: 1)).to eql(['1'])
    end

    it "paginates" do
      stream.vote(field: '3', time: Time.now + one_week)
      expect(stream.get(limit: 1, offset: 0)).to eql([ '3' ])
      expect(stream.get(limit: 1, offset: 1)).to eql([ '2' ])
      expect(stream.get(limit: 1, offset: 2)).to eql([ '1' ])
    end

    it "is possible to get the scores if an option is passed" do
      results = stream.get(with_scores: true)
      results.each do |result|
        expect(result).to be_a(Array)
        expect(result.size).to eql(2)
      end
    end
  end

  describe "#vote" do
    it "ranks most popular for same time as higher" do
      Timecop.freeze do
        stream.vote(field: '1')
        stream.vote(field: '1')
        stream.vote(field: '2')
        expect(stream.get).to eql(['1', '2'])
      end
    end

    it "accepts a weight for the vote of a specific field" do
      Timecop.freeze do
        stream.vote(field: '1')
        stream.vote(field: '1')
        stream.vote(field: '2', weight: 3)
        expect(stream.get).to eql(['2', '1'])
      end
    end

    it "keeps old votes with lower weight on values" do
      Timecop.freeze do
        stream.vote(field: '1', time: Time.now - 2 * one_week)
        stream.vote(field: '2', time: Time.now - 3 * one_week)

        stream.vote(field: '1')
        stream.vote(field: '2')
      end
      expect(stream.get).to eql(['1', '2'])
    end
  end

  it "#count gets the number of elements and #clear! removes them all" do
    stream.vote(field: '1')
    stream.vote(field: '1')
    expect(stream.count).to eql(1)

    stream.clear!
    expect(stream.count).to eql(0)
  end
end


lib/popular_stream.rb
require "popular_stream/version"

# Most of this was taken from:
# http://stdout.heyzap.com/2013/04/08/surfacing-interesting-content/
class PopularStream
  attr_reader :name, :epoch, :max_items

  # 2.5 * half_life (in days) years from now. Make this far in the future!
  DEFAULT_EPOC = Date.new(2017, 5, 29).to_time.to_i
  DAY = 60 * 60 * 24
  HALF_LIFE = 1 * DAY

  class << self
    attr_accessor :redis

    def redis
      @redis ||= Redis.new(url: ENV['REDIS_URL'])
    end
  end

  def initialize(name, **options)
    @name      = name
    @epoch     = options.fetch(:epoch)     { DEFAULT_EPOC }
    @max_items = options.fetch(:max_items) { 10_000 }
  end

  def vote(field:, time: Time.now.to_i, weight: 1)
    time = time.to_i if time.respond_to?(:to_i)

    delta = 2 ** ((time - epoch) / HALF_LIFE)
    redis.zincrby(name, weight * delta.to_f, field)
    trim
  end

  def get(limit: 20, offset: 0, **options)
    redis.zrevrange(name, offset, offset + limit - 1, options)
  end

  def clear!
    redis.del(name)
  end

  def count
    redis.zcard(name)
  end

  def trim
    redis.zremrangebyrank(name, 0, -max_items)
  end

  private

  def redis
    self.class.redis
  end
end
